crates/ruff_python_parser/src/parser/mod.rs

```diff
@@ -0,0 +1,4433 @@
+                    self.source
+                        .len()
+                        .try_into()
+                        .expect("source length is  bigger than u32 max"),
+                ),
+            ),
+            |result| match result {
+                Ok((tok, range)) => (tok.into(), *range),
+                // Return a `Invalid` token when encountering an error
+                Err(err) => (TokenKind::Invalid, err.location),
+            },
+        )
+    }
+
+    #[inline]
+    fn current_token(&mut self) -> (TokenKind, TextRange) {
+        self.lookahead(0)
+    }
+
+    #[inline]
+    fn current_kind(&mut self) -> TokenKind {
+        self.lookahead(0).0
+    }
+
+    fn eat(&mut self, kind: TokenKind) -> bool {
+        if !self.at(kind) {
+            return false;
+        }
+
+        self.next_token();
+        true
+    }
+
+    fn expect(&mut self, expected: TokenKind) -> bool {
+        if self.eat(expected) {
+            return true;
+        }
+
+        let (found, range) = self.current_token();
+        self.add_error(ParseErrorType::ExpectedToken { found, expected }, range);
+        false
+    }
+
+    /// Expects a specific token kind, skipping leading unexpected tokens if needed.
+    fn expect_and_recover(&mut self, expected: TokenKind, recover_set: TokenSet) {
+        if !self.expect(expected) {
+            let expected_set = NEWLINE_EOF_SET
+                .union(recover_set)
+                .union([expected].as_slice().into());
+            // Skip leading unexpected tokens
+            self.skip_until(expected_set);
+
+            self.eat(expected);
+        }
+    }
+
+    fn add_error(&mut self, error: ParseErrorType, range: TextRange) {
+        self.errors.push(ParseError {
+            error,
+            location: range,
+        });
+    }
+
+    fn skip_until(&mut self, token_set: TokenSet) {
+        while !self.at_ts(token_set) {
+            self.next_token();
+        }
+    }
+
+    fn at(&mut self, kind: TokenKind) -> bool {
+        self.current_kind() == kind
+    }
+
+    fn at_ts(&mut self, ts: TokenSet) -> bool {
+        ts.contains(self.current_kind())
+    }
+
+    fn at_expr(&mut self) -> bool {
+        self.at_ts(EXPR_SET)
+    }
+
+    fn at_simple_stmt(&mut self) -> bool {
+        self.at_ts(SIMPLE_STMT_SET2)
+    }
+
+    fn at_compound_stmt(&mut self) -> bool {
+        self.at_ts(COMPOUND_STMT_SET)
+    }
+
+    fn src_text(&self, mut range: TextRange) -> &'src str {
+        // This check is to prevent the parser from panicking when using the
+        // `parse_expression_starts_at` function with an offset bigger than zero.
+        //
+        // The parser assumes that the token's range values are smaller than
+        // the source length. But, with an offset bigger than zero, it can
+        // happen that the token's range values are bigger than the source
+        // length, causing the parser to panic when calling this function
+        // with such ranges.
```

Comment:
The main problem is that `parse_expression_starts_at` passes to the parser a subset of the source code while maintaining the range of the whole source code. If we could pass the entire source code without breaking the behavior of the function, it would solve this problem.